ci-operator/step-registry/cucushift/installer/check/ibmcloud/resourcegroup/cucushift-installer-check-ibmcloud-resourcegroup-commands.sh

```diff
@@ -0,0 +1,96 @@
+#!/bin/bash
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+function ibmcloud_login {
+    export IBMCLOUD_CLI=ibmcloud
+    export IBMCLOUD_HOME=/output
+    region="${LEASED_RESOURCE}"
+    export region
+    "${IBMCLOUD_CLI}" config --check-version=false
+    echo "Try to login..."
+    "${IBMCLOUD_CLI}" login -r ${region} --apikey @"${CLUSTER_PROFILE_DIR}/ibmcloud-api-key"
+}
+
+#OCP-60816 - [IPI-on-IBMCloud] install cluster under BYON with a different resource group
+function getResources() {
+    "${IBMCLOUD_CLI}" resource service-instances --type all --output JSON | jq -r '.[]|.name+" "+.resource_id' | sort
+}
+
+function checkResourcesInVPC() {
+    local preResourceFile provisioned_rg_file resource_group curResourceFile cmpRst
+    resource_group=$1
+    preResourceFile="${SHARED_DIR}/vpc_resources"
+    "${IBMCLOUD_CLI}" target -g ${resource_group}
+    curResourceFile=$(mktemp)
+    getResources > ${curResourceFile}
+    echo "diff ${curResourceFile} ${preResourceFile} ..."
+    cmpRst=$(diff ${curResourceFile} ${preResourceFile})
+    if [ -n "$cmpRst" ]; then
+        echo "ERROR: resources differ: ${cmpRst}"
+        return 1
+    else
+        echo "checkResourcesinVPC PASS"  
+        return 0
+    fi
+}
+
+if [ -f "${SHARED_DIR}/kubeconfig" ] ; then
+    export KUBECONFIG=${SHARED_DIR}/kubeconfig
+else
+    echo "ERROR: fail to get the kubeconfig file under ${SHARED_DIR}!!"
+    exit 1
+fi
+
+if [ -f "${SHARED_DIR}/proxy-conf.sh" ] ; then
+    source "${SHARED_DIR}/proxy-conf.sh"
+fi
+
+check_result=0
+
+provisioned_rg_file="${SHARED_DIR}/ibmcloud_resource_group"
+cluster_resource_group=$(jq -r .ibmcloud.resourceGroupName ${SHARED_DIR}/metadata.json)
+if [ -f "${SHARED_DIR}/ibmcloud_resource_group" ]; then
+    vpc_resource_group=$(cat "${provisioned_rg_file}")
+    rg_file="${SHARED_DIR}/ibmcloud_cluster_resource_group"
+    ibmcloud_login
+    if [ "${CREATE_CLUSTER_RESOURCE_GROUP}" == "yes" ]; then
+        if [ -s "${rg_file}" ]; then            
+            if [ "${vpc_resource_group}" == "${cluster_resource_group}" ]; then
+                echo "ERROR: vpc and cluster use the same resource group ${cluster_resource_group} when CREATE_CLUSTER_RESOURCE_GROUP is set!"
+                check_result=1
+            fi
+            rgName=$(cat "${rg_file}")
+            if [ "${rgName}" != "${cluster_resource_group}" ]; then
+                echo "ERROR: in ${rg_file} saved ${rgName} is not same with the cluster name ${cluster_resource_group}!"
+                check_result=1
+            fi
+
+            if ! checkResourcesInVPC ${vpc_resource_group}; then
+                check_result=1
+            fi
+        else
+            echo "ERROR: fail to find ${rg_file}"
+            check_result=1
+        fi
+    else
+        if [ -s "${rg_file}" ]; then
+            echo "ERROR: find the ${rg_file} when CREATE_CLUSTER_RESOURCE_GROUP is not YES!"
+            check_result=1
+        fi
+    fi
+fi
```

Comment:
The above if logic is not easy to understand, suggest:
```
# checking vpc resource group
provisioned_rg_file="${SHARED_DIR}/ibmcloud_resource_group"
if [ -f "${provisioned_rg_file}" ]; then
    # some codes here to check vpc resources
fi
# checking cluster resource group
if [ "${CREATE_CLUSTER_RESOURCE_GROUP}" == "yes" ]; then
    cluster_resource_group=$(cat "${SHARED_DIR}/ibmcloud_cluster_resource_group")
else
    cluster_resource_group=$(cat "${provisioned_rg_file}")
fi
# The code here to ensure cluster are deployed into cluster_resource_group.
```

Reply:
> cluster_resource_group=$(cat "${provisioned_rg_file}")
need not check whether it is the same with the resource_group in metadata.json

Reply:
from my side, does not need, because we already checked cluster vm is being created in the resource group, checking metadata sounds meaningless.

Reply:
If you really want to check it, let me update the codes with details in a cleaner way:
```
check_result=0
provisioned_resource_group=$(cat "${SHARED_DIR}/ibmcloud_resource_group")

# checking vpc resource group
preResourceFile="${SHARED_DIR}/vpc_resources"
if [ -f "${preResourceFile}" ]; then
    vpc_resource_group="${provisioned_resource_group}"
    checkResourcesInVPC ${vpc_resource_group} || check_result=1
fi

# checking cluster resource group
if [ "${CREATE_CLUSTER_RESOURCE_GROUP}" == "yes" ]; then
    cluster_resource_group=$(cat "${SHARED_DIR}/ibmcloud_cluster_resource_group")
    if [[ "${provisioned_resource_group}" == "${cluster_resource_group}" ]]; then
            echo "ERROR: vpc and cluster use the same resource group ${cluster_resource_group} when CREATE_CLUSTER_RESOURCE_GROUP is set!"
            check_result=1
    fi
else
    cluster_resource_group="${provisioned_resource_group}"
fi
cluster_resource_group_from_installer=$(jq -r .ibmcloud.resourceGroupName ${SHARED_DIR}/metadata.json)
if [[ "${cluster_resource_group_from_installer}" != "${cluster_resource_group}" ]]; then
    echo "ERROR: provisioned cluster resource group does not matched with the one in installer metadata json !"
fi

nodes_count=$(oc get nodes --no-headers | wc -l)
insCount=$(${IBMCLOUD_CLI} is ins --resource-group-name ${cluster_resource_group} --output JSON | jq '.|length')
if [ "${nodes_count}" -ne "${insCount}" ]; then
    echo "ERROR: cluster nodes: $nodes_count; instances in the cluster resource group: $insCount should be same"
    check_result=1
fi
```

Reply:
just when "${CREATE_CLUSTER_RESOURCE_GROUP}" == "yes", need  checkResourcesInVPC

Reply:
`CREATE_CLUSTER_RESOURCE_GROUP` is controlling cluster is installed in the resource group, why need to involve with checking vpc resources?  And if you are concerning if the cluster install deploy some new resource in vpc resource, there is  already a guard - `if [ -f "${preResourceFile}" ]; then`, after all, `checkResourcesInVPC`  is comparing the vpc resources with the previous created ones.

Reply:
checkResourcesInVPC is checking no more cluster resources created in the resource group of VPC. 
when CREATE_CLUSTER_RESOURCE_GROUP is "YES"

Reply:
> checkResourcesInVPC is checking no more cluster resources created in the resource group of VPC.
when CREATE_CLUSTER_RESOURCE_GROUP is "YES"

I already replied that in the last comment, let me copy it here again:
```
And if you are concerning if the cluster install deploy some new resource in vpc resource group, there is already a guard - `if [ -f "${preResourceFile}" ]; then`, after all, `checkResourcesInVPC` is comparing the vpc resources with the previous created ones.
```

Reply:
when CREATE_CLUSTER_RESOURCE_GROUP is not "YES"
checkResourcesInVPC will be failed since all the resources of the cluster will be deployed to one resource group.

Reply:
```
# checking vpc resource group
preResourceFile="${SHARED_DIR}/vpc_resources"
if [ -f "${preResourceFile}" ]; then
    vpc_resource_group="${provisioned_resource_group}"
    checkResourcesInVPC ${vpc_resource_group} || check_result=1
fi
```
How it would be failed? The key point of checking vpc resources or not depends on if you insert `ibmcloud-provision-vpc-resources-info` in your wanted install workflow or not.

Reply:
ibmcloud-provision-vpc-resources-info step will create preResourceFile="${SHARED_DIR}/vpc_resources" file. 
in this step `CREATE_CLUSTER_RESOURCE_GROUP` is `YES` will check `checkResourcesInVPC ${vpc_resource_group}`